lib/shared/option-utils.js

```diff
@@ -0,0 +1,54 @@
+/**
+ * @fileoverview Utilities to operate on option objects.
+ * @author Josh Goldberg
+ */
+
+"use strict";
+
+/**
+ * Determines whether any of input's properties are different
+ * from values that already exist in original.
+ * @template T
+ * @param {Partial<T>} input New value.
+ * @param {T} original Original value.
+ * @returns {boolean} Whether input includes an explicit difference.
+ */
+function containsDifferentProperty(input, original) {
+    if (input === original) {
+        return false;
+    }
+
+    if (Array.isArray(input)) {
+        return (
+            !Array.isArray(original) ||
+            input.length !== original.length ||
+            input.some((value, i) =>
+                containsDifferentProperty(value, original[i]))
+        );
+    }
+
+    if (typeof input !== typeof original) {
+        return true;
+    }
```

Comment:
```suggestion
    if (
        typeof input !== typeof original ||
        Array.isArray(input) !== Array.isArray(original)
    ) {
        return true;
    }

    if (Array.isArray(input)) {
        return (
            input.length !== original.length ||
            input.some((value, i) =>
                containsDifferentProperty(value, original[i]))
        );
    }
```

`containsDifferentProperty({}, [])` would return `false`.